File: leetcode/easy/234-palindromeLinkedList.java
// first pass solution
class Solution {
    public boolean isPalindrome(ListNode head) {
        List<Integer> values = new ArrayList<>();
        
        while (head != null) {
            values.add(head.val);
            head = head.next;
        }
        
        int len = values.size();
        
        for (int i = 0; i < len / 2; i++) {
            if (values.get(i) != values.get(len - i - 1)) {
                return false;
            }
        }
        return true;
    }
}

// second pass solution
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode head2 = reverseList(slow);
        
        while (head != null && head2 != null) {
            if (head.val != head2.val) {
                return false;
            }
            head = head.next;
            head2 = head2.next;
        }
        return true;
    }
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        
        while (curr != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        return prev;
    }
}

File: leetcode/easy/234-palindromeLinkedList.cpp
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* prev = reverseList(slow, nullptr);

        while (prev != nullptr) {
            if (head->val != prev->val) {
                return false;
            }
            head = head->next;
            prev = prev->next;
        }

        return true;
    }
    ListNode* reverseList(ListNode* head, ListNode* prev) {
        if (head == nullptr) return prev;

        ListNode* temp = head->next;
        head->next = prev;
        prev = head;
        head = temp;

        return reverseList(head, prev);  
    }
};
